### PR TITLE
Merge front end plugins feature

### LINKED ISSUE #2084: Front end plugins
## Tasks

Phase 1
- [x] Add a REST endpoint to serve the plugin files #2111
- [x] Add plugin types, create plugin context with hooks #2113
- [x] Implement plugins in the areas defined by the examples #2115
- ~Create a plugin development template #2114~
- [x] Plugin validation - signing and version checks #2164
- [x] Store plugin data #2184 
- [x] Implement plugin events #2186
- [x] Additional framework handling for columns #2187
- [x] Share context with plugins #2148 
- [x] api review feedback #2191 

Phase 2
- Serve plugins configuration from database; load available plugins at server startup
- Create plugin admin page to enable/disable plugins which are currently available
- Sync plugins from the central server

Pase 3
- ~Split common out and publish to it's own package #2100~

### What is a “Plugin” as we’re using the term?
- Discrete code that can be added to a core code base that extends the functionality of the core

### What types of things could a front-end plugin do?
- Add extra fields to an input form with isolated b/e data storage for the plugin
- Add extra validation to input (“you must enter a value between foo and bar in the “Meaningless” field)
- Add toolbar buttons that do special things
- Add a column to a list, and save that data somewhere in an extra field associated with whatever the list is displaying
- Make use of the existing theme for UI consistency
- Integrate with the translation system

### Managing plugins
- Plugins to be loaded as a separate entity to the main code base
- Looking to host them on the server, with the potential for the server to store them in the database and sync the files from central to remote servers
- Have the ability for plugins to be enabled or disabled easily
- Allow enabling via a plugin maintenance screen

### Code isolation
- The ability for plugins to run in an isolated space, and for errors within the plugin to only affect the plugin itself
- All shared interaction between the client application and the plugin is via an interface
- Plugins specify which area they are for, choosing from a pre-defined list. The client app will need to implement plugin support for these pre-defined areas
- Create a plugin interface and class, which is in `common`, with the `common` package published on npm

### Other features
- Allow plugins to use the currently configured language and implement a translation system 
- HRM / react fast reload - can this be available when developing plugins?
- Plugins to be developed using typescript. Make an example / template available
- Versioning and compatibility between f/e version and plugin version. Enforce a dependency version in the plugin
- Plugins can subscribe to events, and raise events? Need to think about data interaction, how data can be passed to a plugin, then retrieved, possibly in a modified form. 

## Implementation

Given the following structures:

```
export enum PluginArea {
  AppBar,
  Toolbar,
}

export enum PluginType {
  InboundShipment,
  InternalOrder,
  OutboundShipment,
  Requisition,
  Stocktake,
}

export interface Plugin<T> {
  area: PluginArea;
  component?: () => React.JSX.Element;
  data?: T;
  name: string;
  path: string;
  type: PluginType;
}

export interface PluginDefinition {
  config: string;
  name: string;
}
```

Looking at adding two REST endpoints to the remote server:

| endpoint | response |
| --------- | --------- |
| `/plugins`   | `PluginDefinition[]` | 
| `/plugins/[name]` | `Plugin` |

With the interaction looking something like this - where the f/e app loads a PluginContext on initial render. The context fetches the plugin definitions from the server and saves them.

When an implementing component is rendered, it requests compatible plugins from the PluginContext and for each one, fetches the code from the server then renders it, passing in the applicable data object. 

[](https://mermaid.live/edit#pako:eNptkU1rwzAMhv-K8GCk0FDGTsthkCbt2GEw1t6aHoSttAZ_4TrbStP_PifOoYUZDJKt531l68K4FcQK1ir7w4_oA2zrxkBcZVY6pyTHIK0BZVGcZtMN5Plrv3CqO0hz6mG5-xzDyppAvwEKSHlNrTRywB8fXp7ift4ngSp7106RJhOSeukJZ6NqIod8Pqlsz456qO89JqH6hhndbs16WP3rk9DV3SMWA5RiMKhp4CEHjkqRgNZ6IORHSBU9rLMv0jYQbMh_k58k1zfd9PC2q6x21kRz8GQEeRJ7NmeavEYp4p9fBqxh4Rg7bFgRQ0Etdio0rDHXWIpdsJuz4awIvqM565zAQLXEg0fNihbVKZ6SkMH6jzTHcZzXP4GGnLQ)

### Toolbar button
Shipping progress plugin for inbound shipments:
- Adds a button to the inbound shipment detail view toolbar
- When clicked it will take the tracking ID from the shipment, connect to an external API and display the shipping progress

### Additional item field
A plugin that will add another field to a stock line:
- Add a column to the stock list to display the additional field
- Add a text input to the stock line edit which has
- Regular expression based validation, and
- Stores the data within the plugin data store, associated with the stock id

### Dashboard widget
A plugin to display data on the dashboard. 
- Q: What would the dashboard data object be like?
- Displays data using standard components. Q: does this require `common` to be public and published?

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2084 

# 👩🏻‍💻 What does this PR do? 
 
Merge the front end plugins work into develop

# 🧪 How has/should this change been tested? 

Check that tests run and the site works. All individual elements have been tested in the PR process.

## 📃 Documentation

- [ ] Plenty of docs to be created

### CODE REVIEW

**Comment by clemens-msupply** [comment]
@mark-prins there are some conflicting files are you working on it? can merge them in...

**Review by clemens-msupply — CHANGES_REQUESTED** [review]
I tried to do a release build and run it. However, the frontend wouldn't load. It says the page js contains some invalid chars... I double-checked with the develop branch and its still working there. I guess it has something to do with the changed webpack config?

Could you please check if it runs for you or if its something on my machine? Hope this is not too difficult to track down...

**Comment by mark-prins** [comment]
> Could you please check if it runs for you or if its something on my machine? 

I've checked out, run `yarn install` and can build the front end just fine. 
I'll ask if someone else can have a try as well. maybe it's linux?

File: server/repository/src/migrations/v1_04_00/mod.rs [hunk 1]
```diff
@@ -5,13 +5,15 @@ pub(crate) struct V1_04_00;
...
 mod date_of_death;
+mod plugin_data;
 
 impl Migration for V1_04_00 {
     fn version(&self) -> Version {
         Version::from_str("1.4.0")
     }
 
     fn migrate(&self, connection: &StorageConnection) -> anyhow::Result<()> {
+        plugin_data::migrate(connection)?;
```
Comment (roxy-dao) [inline server/repository/src/migrations/v1_04_00/mod.rs:16]: move to next version~?

**Comment by mark-prins** [comment]
Killing this one off in favour of #2449

**Pull request closed without merging**